dev/NumberBox/NumberBoxParser.cpp

```diff
@@ -0,0 +1,306 @@
+        {
+            if (m_inputString[m_index + 1] != ' ')
+            {
+                // Parse as a negative number
+                m_index++;
+                ss << m_inputString[m_index];
+            }
+        }
+        else if (PeekNextToken(ExpressionType::Infix).type == MathToken::TokenType::Numeric && m_lastToken.type == MathToken::TokenType::Operator)
+        {
+            // Read next token so that the string is parsed as a numeric token.
+            m_index++;
+            SkipWhiteSpace();
+            ss << m_inputString[m_index];
+        }
+    }
+
+    // Return token that is +,-,*,^,/
+    if (IsOperator(ss.str()))
+    {
+        m_index++;
+        m_lastToken = MathToken(MathToken::TokenType::Operator, ss.str());
+    }
+    // Return parenth token
+    else if (m_inputString[m_index] == '(')
+    {
+        m_index++;
+        m_lastToken = MathToken(MathToken::TokenType::ParenLeft, ss.str());;
+    }
+    else if (m_inputString[m_index] == ')')
+    {
+        m_index++;
+        m_lastToken = MathToken(MathToken::TokenType::ParenRight, ss.str());
+    }
+    // Numeric Tokens
+    else if (IsNumeric(std::wstring_view(ss.str())))
+    {
+        // Keep Reading until all of number is parsed
+        while (IsNumeric(std::wstring_view(ss.str())) && m_index < m_inputLength)
+        {
+            ss << m_inputString[++m_index];
+        }
+        m_lastToken = MathToken(MathToken::TokenType::Numeric, ss.str().substr(0, ss.str().size() - 1));
+    }
+    else
+    {
+        m_index++;
+        m_lastToken = MathToken(MathToken::TokenType::Invalid, ss.str());
+    }
+    return m_lastToken;
+
+}
+
+// Peek at the next token that will be returned. Used for negative checks.
+MathToken MathTokenizer::PeekNextToken(ExpressionType type)
+{
+    // Save statics that need to be restored
+    MathToken oldLastToken = m_lastToken;
+    int oldIndex = m_index++;
+    MathToken nextToken = GetToken(type);
+    // Restore statics
+    m_index = oldIndex;
+    m_lastToken = oldLastToken;
+    return nextToken;
+}
+
+void MathTokenizer::SkipWhiteSpace()
+{
+    while (m_inputString[m_index] == ' ' && m_index <= m_inputLength)
+    {
+        m_index++;
+    }
+}
+
+bool MathTokenizer::IsNumeric(std::wstring_view in)
+{
+    std::wregex r(L"^-?(\\d?)+(\\.)?(\\d?)+?$");
+    return (std::regex_match(in.data(), r));
+}
+
+bool MathTokenizer::IsOperator(std::wstring_view in)
+{
+    if (in.size() > 1)
+        return false;
+    wchar_t op = in.data()[0];
+    switch (op)
+    {
+    case '-':
+    case '+':
+    case '*':
+    case '/':
+    case '^':
+        return true;
+    default:
+        return false;
+    } 
+}
+
+// Determines the mathematical precedence of an operator
+NumberBoxParser::OperatorPrecedence NumberBoxParser::CmpPrecedence(wchar_t op1, wchar_t op2)
+{
+    const std::wstring_view ops = L"-+/*^";
+    int op1prec = (int) ops.find(op1) / 2;
+    int op2prec = (int) ops.find(op2) / 2;
```

Comment:
I really don't like this.. it is dependent on the number of operators and the order that they are declared in the above string. There is definitely a way to do this in a type safe manor